### ISSUE TITLE
Outdated OpenSSL and hyper-tls 

### ISSUE THREAD

**Issue by richardwhiuk** [Metaswitch/swagger-rs#66]
As per @CodeSandwich in #51 

> I'd suggest upgrading yet another dependency, which is burried deep in other outdated crates: openssl. Current version (0.9.24) does not compile on systems with OpenSSL 1.1.1.

Sadly we can't update OpenSSL to a later version.

In native-tls 0.2, which supports Rust OpenSSL 0.10, they've dropped support for `TlsConnectorBuilderExt::from_openssl`, which we use in https://github.com/Metaswitch/swagger-rs/blob/f86922a3b9a6967831faa891922f0bedf38d0fec/src/connector.rs#L88-L99

Until we add sufficient function to native TLS to do what we are doing by using OpenSSL manually, we can't move.

I've raised https://github.com/sfackler/rust-native-tls/issues/125 to look at getting a fix to unblock this - until then, this issue is blocked.

**Comment by richardwhiuk** [comment]
Did some more debugging as to what a fix here might look like:

`swagger-rs` uses `hyper-tls` to do TLS, which requires providing a `native_tls::TlsConnectorBuilder`.

`native-tls~0.1` supports `openssl~0.9`
`native-tls~0.2` supports `openssl~0.10`

`native-tls~0.1` supports creating a `native_tls::TlsConnectorBuilder` by wrapping a `openssl::SslConnector`.
`native-tls~0.2` doesn't support this - it requires everything to be done in ways that work cross platform (SChannel on Windows, Secure Transport on OS X, OpenSSL on all other platforms)

`openssl::SslConnector` allows you to do the following things which `swagger` does.
- Set the Root Certificates (as opposed to adding Root Certificates alongside the system trust store)
- Set a Private Key and Certificate for mutual TLS by using a PEM file

Neither of those is currently supported by `native_tls::TlsConnectorBuilder`.

The first of those actually looks impossible to achieve directly in Windows SChannels - in particular CertGetCertificateChain - https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certgetcertificatechain - allows you to provide an additional certificate store, but not a default one.

Next step was to verify that the current behavior of Swagger is that it will only allow access via the provided root certificates (ignoring the system store).

If so, I think the enhancement to schannel would need to include the ability to spot that the provided chain didn't include one of the root certificates in the provided additional store, which is suboptimal. We'd need to ensure we asked for all chains to avoid a double chaining problem not causing problems.

**Comment by richardwhiuk** [comment]
New plan:
- hyper-openssl exists, and we can replace hyper-tls with that.

That doesn't solve cross platform support, but will solve this issue, and https://github.com/OpenAPITools/openapi-generator/issues/3664, which is probably more pressing.

We'll also do better on platforms which don't support OpenSSL, and not even generate panic-ing code.

That will all be a breaking change to swagger, and debatabley the code-genned code.

**Comment by larstobi** [comment]
I've tracked down this issue from a problem with building the "rust-server" code generator from the openapi-generator tool. It uses swagger = "2", however, even if I update it to the latest version swagger = "3.1.0" it still bugs out while building openssl v0.9.24, which fails since I have OpenSSL 1.1.1.

@richardwhiuk is there any chance we can get PR #79 merged?

**Comment by richardwhiuk** [comment]
This is fixed in Swagger 4.0, which is now available on crates.io
